nix/test-environments/test-script.py

```diff
@@ -0,0 +1,300 @@
+import re
+import shlex
+import json
+import time
+from dataclasses import dataclass
+from typing import List
+from test_driver.errors import RequestedAssertionFailed # type: ignore
+
+# --- Configuration ---
+@dataclass
+class TestConfig:
+    """Centralized configuration for the integration test."""
+    BASE_URL: str = "http://127.0.0.1"
+    INK_PORT: int = 8547
+    SEPOLIA_PORT: int = 8546
+    SEQUENCER_PORT: int = 5553
+
+    # Contract Addresses
+    UPGRADEABLE_PROXY_CONTRACT: str = "0xee5a4826068c5326a7f06fd6c7cbf816f096846c"
+    CHAINLINK_PROXY_CONTRACT: str = "0x9fAb38E38d526c6ba82879c6bDe1c4Fd73378f17"
+    UPGRADEABLE_PROXY_ADFS_CONTRACT: str = "0xADF5aad6faA8f2bFD388D38434Fc45625Ebd9d3b"
+    CL_AGGREGATOR_ADAPTER_CONTRACT: str = "0xcBD6FC059bEDc859d43994F5C221d96F9eD5340f"
+
+    # Service Names
+    REPORTER_SERVICE: str = "blocksense-reporter-a.service"
+    ANVIL_SEPOLIA_SERVICE: str = "blocksense-anvil-ethereum-sepolia.service"
+    ANVIL_INK_SERVICE: str = "blocksense-anvil-ink-sepolia.service"
+    SEQUENCER_SERVICE: str = "blocksense-sequencer.service"
+
+    # Timeouts
+    ENDPOINT_EXISTENCE_TIMEOUT: int = 900
+    HISTORY_POPULATION_TIMEOUT: int = 360
+    NETWORK_UPDATE_TIMEOUT: int = 180
+    VALUE_CHANGE_TIMEOUT: int = 20
+    VALUE_CHANGE_POLL_INTERVAL: int = 1
+
+    @property
+    def INK_RPC_URL(self) -> str:
+        return f"{self.BASE_URL}:{self.INK_PORT}"
+
+    @property
+    def SEPOLIA_RPC_URL(self) -> str:
+        return f"{self.BASE_URL}:{self.SEPOLIA_PORT}"
+
+    @property
+    def SEQUENCER_URL(self) -> str:
+        return f"{self.BASE_URL}:{self.SEQUENCER_PORT}"
+
+    @property
+    def services_to_wait_for(self) -> List[str]:
+        return [
+            self.ANVIL_SEPOLIA_SERVICE,
+            self.ANVIL_INK_SERVICE,
+            self.SEQUENCER_SERVICE,
+            self.REPORTER_SERVICE,
+        ]
+
+    @property
+    def ports_to_wait_for(self) -> List[int]:
+        return [self.SEPOLIA_PORT, self.INK_PORT, self.SEQUENCER_PORT]
+
+# Create a single instance of the config to be used throughout the script
+test_config = TestConfig()
+
+
+def wait_for_services(services):
+    for service in services:
+        machine.wait_for_unit(service)  # type: ignore
```

Comment:
Global usage of `machine` seems icky, could we maybe substitute it for a more descriptive name? Maybe also somehow, somewhere declare it's actual type, so that the automatic type-checker does not yell at us?